Update flow go crypto version

tarakby opened the pull request:
Update the crypto version in Flow-go and integrate the new breaking changes of the crypto module. In particular:
 - use the new sentinel errors of the crypto module (empty list to aggregate, non BLS key, signature that doesn't serialize a valid point), mainly in consensus and fvm.
 - add new signature module error for identity aggregated public key: `Aggregate()` returns the new sentinel `IdentityPublicKeyError` when the aggregated public key is identity. verifying any signature with the identity key results in `false` in the new crypto version, and hence no signature should be returned by `Aggregate()` when key is identity.
 - update consensus/hotstuff: a leader that aggregates signatures checks the aggregated public key is not identity. This is a sanity check because the identity public key can happen in 2 unlikely cases:
      - hotstuff quorum has at least one honest actor: it's not possible for remaining malicious actors to submit public keys that add up to identity, given they have to submit valid corresponding PoPs. 
      - all hotstuff quorum is malicious: this is unlikely by the overall protocol assumptions.
 - `Aggregate()` function for TC construction isn't updated for now because signature verification under multiple messages isn't impacted by the identity public key constraint (according to the IETF BLS standard draft) - though I think the draft has overlooked some edge cases (I reached out to the standard authors to clarify and will give an update later).

tarakby commented on consensus/hotstuff/model/errors.go:
```diff
@@ -233,6 +233,18 @@ func IsInvalidSignatureIncludedError(err error) bool {
 	return errors.As(err, &e)
 }
 
+// InvalidAggregatedSignatureError indicates that the aggregated signature is invalid.
```
consensus/hotstuff/model and module/signature packages have been using locally defined errors already, that may seem like duplicates. I believe this was a choice made by the consensus team. You can see that for instance with `InsufficientSignaturesError` or `InvalidSignatureIncludedError` that are defined by both packages.

In this PR, I continued using that convention for the newly added `InvalidAggregatedSignatureError`. 
cc @AlexHentschel for visibility.

AlexHentschel replied:
While embedded into the code base, we are trying to implement the consensus such that it is largely decoupled from the remaining code base. This is because we are eventually hoping to isolate the consensus, so academic researchers can run it by itself without needing to spin up a full flow network. Thereby it would become much easier to invite academic contributions, if researchers could directly work on the code base and use it for their consensus experiments. 

Therefore, we are trying to not expose any lower-level errors to the consensus system and instead have a clean abstraction hiding any lower-level platform code such as cryptographic details.

tarakby commented on consensus/hotstuff/signature/randombeacon_inspector.go:
```diff
@@ -33,7 +33,7 @@ func NewRandomBeaconInspector(
 		signature.RandomBeaconTag)
 	if err != nil {
-		if crypto.IsInvalidInputsError(err) {
+		if crypto.IsInvalidInputsError(err) || crypto.IsNotBLSKeyError(err) {
```
As mentioned above, the new crypto version comes with breaking changes. In particular, new crypto specific sentinel errors have been defined, and are replacing the generic `IsInvalidInputsError` (more details this PR description https://github.com/onflow/flow-go/pull/2861). `IsInvalidInputsError` is still used for other error cases for which sentinels are not defined. 

The crypto functions now can return multiple errors depending on the specific invalid behaviour. Not all functions return the same sentinel errors and that's why the generic `IsInvalidInputsError` is replaced with different checks, depending on the function called and the context. The crypto public function docs list the cases and errors returned and should be checked by developers.
As the FVM codeowner, maybe you could validate the error changes in the fvm file.

pattyshack replied:
I don't see why this needs to be a breaking change.

InvalidInputError can and should internally call the new predicates since these are also input errors.

anyways.  i've unblock my request for change.

AlexHentschel replied:
There is a longer back-story to this. In a nutshell, we realized that using only `InvalidInputsError` is too coarse and makes it difficult for higher-level code to differentiate internal implementation bugs from byzantine inputs.

I agree that we could still keep the `InvalidInputsError` as an umbrella error, which wraps more specific error cases. But the code interacting with the lower-level crypto implementations is actually not that much. Then the question becomes: it it worth-while to expose an error hierarchy through the API, or is it maybe less work and less complexity to just do a breaking change? It is not really apparent to me, which is less work-intensive. But I feel Tarak's choice to go with the breaking change is certainly a viable approach.

tarakby replied:
From the crypto module perspective, these are all invalid inputs error. From the Flow protocol perspective, these errors can be triggered by different scenarios, and the node has to react differently depending on the scenario detected (as part of the BFT milestone). The sentinel errors break down the original `InvalidInputsError` to help the node identify the issue (malicious behaviour detected/unexpected error..). 
The breaking changes in the crypto module force the protocol code to be updated and to not rely only on the generic inputs error.

tarakby replied:
From the crypto module perspective, these are all invalid inputs error. From the Flow protocol perspective, these errors can be triggered by different scenarios, and the sentinel errors break down the original InvalidInputsError to help the node identify the issue.

Initially, I used an "umbrella error" and sub-error types to avoid the breaking change. That meant the higher layer would need to call multiple embedded error checks to identify the issue. I have finally opted for simpler errors to simplify the checks on the higher layer. Moreover, the breaking change forces integrating the new crypto module version to look at all cases carefully, instead of relying on the "umbrella error" and missing some important sub-cases.

AlexHentschel reviewed (commented):
Overall, it looks good. There are a variety of design aspects that I think could be polished a bit more. I created PR https://github.com/onflow/flow-go/pull/3677 targeting your branch to illustrate my suggested revisions (and fixing some typos which I randomly noticed). Most of my comments are covered by my PR. 

There is **one more aspect** (not covered in my PR), which I was hoping to discuss further:
* If I understand this documentation correctly, there is some overlap between `InvalidAggregatedSignatureError` and `InvalidSignatureIncludedError`, which the current implementation is unable to differentiate: https://github.com/onflow/flow-go/blob/67c382b439ede161df93b2496e6fd8efe9366b6b/module/signature/aggregation.go#L174-L177
* I wonder if we can resolve this by 
   1. checking whether the aggregated key is the identity key and erroring in this case here: https://github.com/onflow/flow-go/blob/67c382b439ede161df93b2496e6fd8efe9366b6b/module/signature/aggregation.go#L243 I am hoping that this would cover the case, where the keys were forged to produce the identity signature.
   2. In case the aggregated key is _not_ the identity key, we could conclude that therefore some signatures must be invalid, because otherwise, the aggregated signature should also yield the identity signature (is this a correct assumption?).

The reason why I think it would be valuable to differentiate these two cases is:
* we could differentiate the case where there is a bug in the node vs the case where the node is operating in a majority byzantine environment. While this is a very strong edge case, it would still be nice to be able to draw this conclusion (if possible with reasonable engineering effort).

tarakby commented on consensus/hotstuff/model/errors.go:
```diff
@@ -233,6 +233,18 @@ func IsInvalidSignatureIncludedError(err error) bool {
+// This can happen because:
+//   - one or many signatures added via TrustedAdd are invalid to their respective public keys.
+//   - OR the signatures are valid but the public keys were forged to sum up to an identity public key.
+var InvalidAggregatedSignatureError = errors.New("aggregated signature is invalid (identity signature)")
```
I agree with the naming/definition of error proposed in https://github.com/onflow/flow-go/pull/3677

tarakby commented on consensus/hotstuff/verification/combined_verifier_v2.go:
```diff
@@ -147,17 +148,21 @@ func (c *CombinedVerifier) VerifyQC(signers flow.IdentityList, sigData []byte, b
-		// Hence, if we encounter an error here, we know it is case (ii). Thereby, we can clearly
-		// distinguish a faulty _external_ input from an _internal_ uncovered edge-case.
+		// check case (i)
+		// TODO: this should not happen because the case `len(signers) == 0`
+		// is checked upfront and ruled out. This is because `Unpack` does not
+		// handle the empty list case properly.
```
> Do you want to indicate that in the future we should make the following revisions?

Yes, that was the meaning of the TODO. 

My bad, I initially thought the tests were failing because of `Unpack`, but they were failing because of mockery issues. I did fix the mocks in tests but didn't get back to testing `Unpack` with an empty list. 

Thank you for cleaning the checks in https://github.com/onflow/flow-go/pull/3677. The TODOs are now not needed.

tarakby replied:
I actually dug further, and it turns out the empty list case was failing because the test itself. `Unpack` was failing because the test was not built properly I think. Fixed in [e5e3930](https://github.com/onflow/flow-go/pull/3441/commits/e5e3930ff0f9c2456829b221c95c2af821bcfc3f)

tarakby commented:
> If I understand this documentation correctly, there is some overlap between InvalidAggregatedSignatureError and InvalidSignatureIncludedError, which the current implementation is unable to differentiate

Yes, there is an overlap indeed, and you are right, comparing the aggregated public key is enough to clear up the overlap. 
I initially implemented the public key check, but then decided to switch to the signature check, for the following reasons:

- the current implementation uses `VerifyAggregate` for the final verification in `Aggregate`. The aggregate public key computation is abstracted under `VerifyAggregate`. If we were to compute the aggregate public key within `Aggregate`, we would break the nice modularity of the code. 
- is differentiating the overlap important? : whether the public keys were constructed to yield to identity key, OR whether the received votes were constructed to yield to identity signature, the node would conclude that it is operating in a majority malicious environment either way. If some signatures are invalid, the node would be able to detect it through individual verifications.
- In the way we are currently using the aggregation module in Hotstuff, `TrustedAdd` is always called after a successful `Verify`. If `Aggregate` is called by a Hotstuff leader, it is guaranteed that all signatures have been validated individually. If we fall under the identity signature case, it is guaranteed that we are in the case were public keys were maliciously crafted to yield to an identity key.

tarakby commented:
@AlexHentschel PR should be ready for another round:
 - incorporated changes from https://github.com/onflow/flow-go/pull/3677
 - updated identity checks to target aggregated public key instead of aggregated signature. 
 - renamed `module/signature.ErrIdentitySignature` to `module/signature.ErrIdentityPublicKey` to match the specific error logic
 - did not rename the higher level `hotstuff/model.InvalidAggregatedSignatureError`. Technically the reason `Aggregate()` function in this module errors because the aggregated signature would be invalid.

AlexHentschel commented on module/signature/aggregation.go:
```diff
@@ -349,7 +360,7 @@ func (p *PublicKeyAggregator) KeyAggregate(signers []int) (crypto.PublicKey, err
 		updatedKey, err = crypto.RemoveBLSPublicKeys(updatedKey, missingSignerKeys)
 		if err != nil {
-			// not expected in notrmal operations as there is at least one key, and all keys are BLS
+			// not expected in normal operations as there is at least one key, and all keys are BLS
```
```suggestion
			// no error expected as there is at least one key (checked above), and all keys are BLS (checked in the constructor)
```

same suggestion applies to [line 357](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation.go#L357) (still with the typo "notrmal"): https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation.go#L357

AlexHentschel commented on module/signature/aggregation_test.go:
```diff
@@ -218,6 +230,34 @@ func TestAggregatorSameMessage(t *testing.T) {
 		}
 	})
 
+	t.Run("identity aggregated signature", func(t *testing.T) {
```
I like how you describe the different failure cases here: https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation_test.go#L191-L195

but I have trouble matching the test cases [`invalid signature`](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation_test.go#L196) and [`identity aggregated signature`](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation_test.go#L233) to the description

_Suggestion_
* My understanding is that failure case 1 and 3 are what is tested in [`invalid signature`](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation_test.go#L196) (?) If that is the case, maybe we can revise the description to focus on the expected sentinel. For test  [`invalid signature`](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation_test.go#L196), we could revise test [`invalid signature`](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation_test.go#L190-L199) to 
    ```golang
    // There are two scenarios, where `Aggregate()` is expected to return an `InvalidSignatureIncludedError`:
    //  1. Some signature(s) added via `TrustedAdd` has an invalid structure, i.e. it can't be deserialized
    //     successfully. In this case, already the aggregation step fails.
    //  2. The signatures were deserialized successfully, but the aggregate signature doesn't verify to the
    //     aggregate public key (although it is not identity).
    t.Run("invalid signature", func(t *testing.T) {
      _, s, _ := createAggregationData(t, 1)
      invalidStructureSig := (crypto.Signature)([]byte{0, 0}) // case 1: signature with invalid structure
      mismatchingSig := s[0]                                  // case 2: structurally valid sig, but doesn't match respective key
         ⋮
    ```
* For test [`identity aggregated signature`](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation_test.go#L233):
   * would suggest to update the documentation 
      ```golang
      // `Aggregate()` is expected to return an `ErrIdentityPublicKey` when the aggregated signature is identity signature;
      //  even if the signatures are all individually valid
      t.Run("identity aggregated signature", func(t *testing.T) {
         ⋮
      ```
   * Same recommendation as in [my comment above](): minimal test cases without any overlap with the [`invalid signature` test case](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/module/signature/aggregation_test.go#L196); in other words, lets make sure `ErrIdentityPublicKey` is returned even if all signatures are individually valid.

tarakby replied:
> I like how you describe the different failure cases here

I think this was your description from your suggestion PR in the previous iteration :) 
I have updated the comment because we shifted the logic from checking for identity signatures to checking for identity public key. Please refer to this comment for more details : https://github.com/onflow/flow-go/pull/3441#discussion_r1095435637

AlexHentschel reviewed (approved):
Awesome PR. I added a lot of comments, but they are mostly suggestions for revising documentations. I tried to add wording suggestions, which you could just directly apply or copy if you like them.

tarakby commented on module/signature/aggregation.go:
```diff
@@ -161,16 +161,21 @@ func (s *SignatureAggregatorSameMessage) HasSignature(signer int) (bool, error)
+//   - ErrIdentityPublicKey if the signer's public keys add up to the BLS identity public key.
+//     The aggregated signature would fail the cryptographic verification if verified against the
+//     the identity public key. This case can only happen if public keys were forged to sum up to
+//     an identity public key.
```
My intention was to make `module/signature` a generic cryptographic tool that is abstracted from the possible applications. In that sense, an identity public key is a possible case for the aggregated key but very unlikely if all keys are uniformly sampled. `consensus/hotstuff/signature` is the application and that's when I think the term "malicious" makes sense. That has been described [here](https://github.com/onflow/flow-go/blob/f0bbe6b6b05003dad0efdb900d9336c582d8afbf/consensus/hotstuff/signature/weighted_signature_aggregator.go#L156-L161). 

Your comment still reminded me to add further precision in both modules 🙏🏼 I added it here : https://github.com/onflow/flow-go/pull/3441/commits/0706067b10866275cabbdb2ae8d59d7d1783b26e

tarakby commented on fvm/crypto/crypto.go:
```diff
@@ -303,7 +303,7 @@ func AggregatePublicKeys(keys []*runtime.PublicKey) (*runtime.PublicKey, error)
+		if crypto.IsBLSAggregateEmptyListError(err) || crypto.IsNotBLSKeyError(err) {
 			return nil, err
 		}
 		panic(fmt.Errorf("aggregate BLS public keys failed with unexpected error %w", err))
```
First, I think there is a little typo in the commented function (suggested code should be about `AggregatePublicKeys`). 

Then, I'm not sure if there is still a TODO for this one. This is FVM code that is called by Cadence runtime. My understanding from discussing with the Cadence/execution side is that only errors caused by the Cadence users should be "bubbled up" and returned to Cadence. All other unexpected errors should panic.
In this case (`AggregatePublicKeys`), the user errors are:
 - the user uses non BLS keys
 - the user attempts to aggregate an empty list of keys. 
 Both errors are handled, everything else is supposed to panic (all other errors [are unexpected anyway](https://github.com/onflow/flow-go/blob/bdba0b4fc9d25113d5f6650edc15b0561c2c4f89/crypto/bls_multisig.go#L175-L178)).  The case of `AggregateSignatures` is similar.
 I might have missed something.

tarakby replied:
The current pattern was advised by the execution team long time ago, a `panic()` call in the crypto implementation doesn't cause the node to crash, it is handled by FVM and only the transaction fails. There is an ongoing discussion with the execution team to use a cleaner pattern.

tarakby replied:
The current pattern was confirmed after a discussion with the execution team.

tarakby commented on model/convert/service_event.go:
```diff
@@ -395,9 +395,25 @@ func convertClusterQCVotes(cdcClusterQCs []cadence.Value) ([]flow.ClusterQCVoteD
+		//  - If the quorum has at least one honest signer, the aggregated public is uniformly sampled
+		//    and the identity key probability is negligible.
+		//  - If all quorum is malicious and intentionally forge an identity aggregate. This is also
+		//    unlikely since the clusters are proven with high probability not to have a malicious quorum.
```
Yes, what I meant is a "full malicious quorum" i.e a quorum with strictly more than 2/3 of the stake and with only malicious nodes. My understanding is that this is proven very unlikely given the the overall 1/3 malicious fraction upper-bound axiom. 

Updated in https://github.com/onflow/flow-go/pull/3441/commits/e398df5de1b66cfd656caa9a36dd7b61916cc51f. 
Note that I updated my wording of the first case too, as I noticed a subtle mistake. The first case (at least one honest node) has two sub-cases:
  - all nodes are honest. In this case the aggregated public key is indeed uniformly sampled. 
  - at least one remaining node is malicious. The aggregated key in this case isn't necessary uniform (when malicious keys are submitted after the honest key is submitted). However, it's not computationally possible for malicious nodes to target the identity key (thanks to PoPs!).
  
I found this a bit too detailed, and merging the two cases somehow made sense to me. but I'm happy to add the sub-cases if you think it's relevant.

tarakby commented on module/signature/aggregation.go:
```diff
@@ -161,16 +161,21 @@ func (s *SignatureAggregatorSameMessage) HasSignature(signer int) (bool, error)
+// It also errors if no signatures were added.
 // Post-check of aggregated signature is required for function safety, as `TrustedAdd` allows
-// adding invalid signatures. The function is not thread-safe.
+// adding invalid signatures. Aggregation may also output an invalid signature (identity)
+// even though all included signatures are valid (extremely unlikely case when all keys are sampled
+// uniformly)
```
my bad, we have spoken about a similar case in the past. The update makes sense

tarakby commented on consensus/hotstuff/signature/weighted_signature_aggregator_test.go:
```diff
@@ -238,4 +238,30 @@ func TestWeightedSignatureAggregator(t *testing.T) {
+		require.NoError(t, err)
+
+		// add invalid signature for signer with index 1
+		// (invalid because the corresponding key was altered)
```
Thanks for pointing this out. I agree about having tests with minimal causes of errors.
I also think it makes sense in some subtle cases to test the black box implementation returns the expected sentinel error when 2 causes for errors overlap ("expected" means defined by the application). 

In our context, notice that we have updated the logic of building a QC to not rely on the identity signature, but to rely on the identity public key instead. This allows the higher logic to isolate the case of a colluding committee (when signature is identity, we're not sure whether some signatures are invalid or if all signatures are valid and key is identity).
I think the test cases (1) and (2) above do not reflect the expected behaviour because:
    - there should be case (3) for completeness where: the resulting aggregate is the identity signature and some signatures are invalid
    - in cases (1) and (3) there is an overlap that we want to distinguish. Whether the aggregated key is identity or not results in a different sentinel error.


I rather see the unhappy path test cases from the application perspective (for a black box implementation) as:

 I. aggregated key is not identity but some individual signatures are invalid (minimal error) : `model.InvalidSignatureIncludedError`  
  (already [implemented](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/consensus/hotstuff/signature/weighted_signature_aggregator_test.go#L194))
 II. aggregated key is identity and all individual signatures are valid (minimal error) : `model.InvalidAggregatedSignatureError` (missing test!)
 III. aggregated key is identity and some individual signatures are invalid (overlapping errors): `model.InvalidAggregatedSignatureError` ([implemented](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/consensus/hotstuff/signature/weighted_signature_aggregator_test.go#L241) but the test name was not updated from the previous iteration and hence the confusion!)
 
 For each case (I) (II) and (III), we could add sub tests for (a) identity and (b) non-identity aggregated signatures for better robustness, but that's not relevant for the application logic at this point.

In the above, "invalid signatures" mean they are correctly serialized but do not verify against the public key. 
However I noticed a subtle case when at least a signature does not deserialize correctly:
  - if aggregated key is non-identity, expect `model.InvalidSignatureIncludedError` (minimal error) (already implemented)
  - if aggregated key is identity, expect `model.InvalidSignatureIncludedError` (overlapping error). Why not `model.InvalidAggregatedSignatureError` like the overlapping case above? I think here we didn't reach the cryptographic validity check yet, and we are still checking the collected data prior to the aggregation work. So it makes sense to not speak about an invalid aggregated signature. Happy to hear your thought. 
  
I have implemented the changes in https://github.com/onflow/flow-go/pull/3441/commits/31532375009b84b3cb2afec59a9579b24939079a

tarakby replied:
Fyi, I have updated your suggestion to make `createAggregationData` not instantiate the aggregator (though returning `sks` isn't required here).

tarakby replied:
I've updated the expected returned errors of `Aggregate()` docs in https://github.com/onflow/flow-go/pull/3441/commits/22a585458a9926a7eb18c6de30738211ffbf8c99

AlexHentschel replied:
> if aggregated key is identity, expect `model.InvalidSignatureIncludedError` (overlapping error). Why not `model.IsInvalidAggregatedKeyError` like the overlapping case above? I think here we didn't reach the cryptographic validity check yet, and we are still checking the collected data prior to the aggregation work. So it makes sense to not speak about an invalid aggregated key. Happy to hear your thought.

This is a relatively broad topic. Imagine a function that can return 10 different sentinels. How do we expect the function to react to an input that violates all 10 correctness conditions. Documenting precedence of failure conditions is generally not tractable in that it requires a lot more work and often implementation details need to be exposed to explain the mechanics. 
I like to treat sentinel errors as non-deterministic, where reporting one violation of happy path is sufficient to reject the input. We do no need to (and often can't) return the exhaustive set of _all_ contract violations in the input. 

So in general, I would say the method can just non-deterministically return any one applicable sentinel.

In this particular case, you are making the argument that the more detailed differentiation is useful for the use-case (?) If it is not too much more work, by all means, with a clean documentation, we can explain the precedence of the different failure cases.

I am just saying, we don't have to ... and there does _not_ have to be any particular logic as to why sentinel `errA` is returns over `errB`.

AlexHentschel replied:
thank you, current implementation and test look great.

tarakby commented on consensus/hotstuff/verification/combined_verifier_v3.go:
```diff
@@ -153,10 +151,24 @@ func (c *CombinedVerifierV3) VerifyQC(signers flow.IdentityList, sigData []byte,
 	// Caution: this function will error if pubKeys is empty
 	verifyAggregatedSignature := func(pubKeys []crypto.PublicKey, aggregatedSig crypto.Signature, hasher hash.Hasher) error {
```
This is redundant code indeed, though I didn't want to refactor the existing Hotstuff code (kinda unrelated to the crypto version integration). This refactor makes the code better, I implemented it your suggestion with minor changes in https://github.com/onflow/flow-go/pull/3441/commits/25befa3c3c645a4a3b85a93f82945319a113d657

tarakby replied:
Note that `blockID` was taken out of the function because it is already logged by the calling code.

AlexHentschel commented on consensus/hotstuff/signature/weighted_signature_aggregator_test.go:
```diff
@@ -238,4 +242,174 @@ func TestWeightedSignatureAggregator(t *testing.T) {
+			// Aggregation should error with sentinel InvalidSignatureIncludedError
+			// aggregated public key is identity
+			signers, agg, err := aggregator.Aggregate()
+			assert.True(t, model.IsInvalidSignatureIncludedError(err))
```
I think we could allow here to return either `InvalidSignatureIncludedError` or  `InvalidAggregatedKeyError`

AlexHentschel commented on consensus/hotstuff/signature/weighted_signature_aggregator_test.go:
```diff
@@ -238,4 +242,174 @@ func TestWeightedSignatureAggregator(t *testing.T) {
+			// aggregated public key is identity
+			signers, agg, err := aggregator.Aggregate()
+			assert.Error(t, err)
+			assert.True(t, model.IsInvalidAggregatedKeyError(err))
```
same here, the test could pass if either `InvalidSignatureIncludedError` or  `InvalidAggregatedKeyError` is returned

AlexHentschel commented on consensus/hotstuff/verification/common.go:
```diff
@@ -27,3 +32,104 @@ func MakeTimeoutMessage(view uint64, newestQCView uint64) []byte {
+	sigData crypto.Signature,
+	view uint64,
+	highQCViews []uint64,
+	hasher hash.Hasher) error {
```
+1

AlexHentschel commented on module/signature/aggregation.go:
```diff
@@ -213,41 +230,40 @@ func (s *SignatureAggregatorSameMessage) Aggregate() ([]int, crypto.Signature, e
-	if sharesNum == 0 {
-		return false, NewInsufficientSignaturesErrorf("cannot aggregate an empty list of signatures")
-	}
+func (s *SignatureAggregatorSameMessage) VerifyAggregate(signers []int, sig crypto.Signature) (bool, crypto.PublicKey, error) {
```
do you have an opinion on moving the [check for identity public key](https://github.com/onflow/flow-go/blob/11d919c379660a1e866cf4b1eb3bc06c1a1f0cff/module/signature/aggregation.go#L219-L221) into this function? 

In essence, that would mean that the boolean return will only be false, if one of the input signatures is invalid. The case, where the aggregated key is identity is conceptually a setup error, where the `SignatureAggregatorSameMessage` has technically a dysfunctional configuration. From that angle it would make sense to handle these two scenarios differently on the API. 

I feel this would make the API a bit more expressive, and simpler 🤷 Alternatively, we could add a couple sentences to the method's goDoc explaining what a `false` return value implies and how to differentiate the two cases. (essentially describing [this](https://github.com/onflow/flow-go/blob/11d919c379660a1e866cf4b1eb3bc06c1a1f0cff/module/signature/aggregation.go#L217-L225))

tarakby replied:
I gave the suggestion thoughts and I prefer the current "logic" of outputs, although I agree it simplifies the code a bit:

- IMHO, an error should result from a dysfunctional configuration or state. Verifying a signature against an identity key isn't a dysfunctional configuration. It is a normal setup that results in the function answering `false` to the question "is there enough guarantee that the private key corresponding to the input public key generated this input signature using this input message". If the signature has an invalid length or serialization, the answer to the question is `false` but no error is triggered. In the case of the identity public key, the answer is also `false` because the valid signature is the same for any message (i.e we can't be sure the private key has signed that specific input message), but there is no invalid setup of `SignatureAggregatorSameMessage`.  Therefore I don't fully agree to make `VerifyAggregate` return an error for this case. 
- Why would `Aggregate` return an error in this case? An aggregator intends to generate an aggregated signature that would eventually be valid when verified. If the aggregator fails to generate such valid signature, it makes sense for it to error.
- how would the caller of `VerifyAggregate` differentiate between an invalid signature because of invalid individual signature and identity public key? I think the function should provide extra data to differentiate those cases without errors, in this case I chose the aggregated key itself (but a simple boolean also works). I chose to output the key in case some future uses find it useful. 

I will add extra documentation for `VerifyAggregate` to clarify how to differentiate the subcases.

tarakby replied:
> that would mean that the boolean return will only be false, if one of the input signatures is invalid

I also wanted to add that from the perspective of `VerifyAggregate`, there are no individual signatures. The function takes an input signature to verify, it doesn't consider any saved individual signature from `SignatureAggregatorSameMessage`, and ignores how that input signature was generated. The stateful object is only used to read the public keys we would like to test against.

tarakby replied:
https://github.com/onflow/flow-go/pull/3441/commits/c192f795e1a6d214c7dfd1383c81c3a7727d4c71

AlexHentschel reviewed (approved):
Very nice. This can be merged. All my comments are optional: largely aesthetic, and polishing of godoc.

AlexHentschel commented on model/convert/service_event.go:
```diff
@@ -395,9 +395,34 @@ func convertClusterQCVotes(cdcClusterQCs []cadence.Value) ([]flow.ClusterQCVoteD
+		//  - If the quorum has at least one honest signer, and given all staking key proofs of possession
+		//    are valid, it's extremely unlikely for the aggregated public key (and the corresponding
+		//    aggregated signature) to be identity.
+		//  - If all quorum is malicious and intentionally forge an identity aggregate. As of the previous point,
+		//    this is only possible if there is no honest collector involved in constructing the cluster QC.
+		//    Hence, the cluster would need to contain a supermajority of malicious collectors.
+		//    As we are assuming that the fraction of malicious collectors overall does not exceed 1/3  (measured
+		//    by stake), the probability for randomly assigning 2/3 or more byzantine collectors to a single cluster
+		//    vanishes (provided a sufficiently high collector count in total).
```
> In general the protocol needs to handle a Byzantine cluster.

I agree, but this is a very extreme edge case. Assuming close to 1/3 byzantine fraction in the overall collector population, it is reasonably likely to sample a cluster, whose local fraction of byzantine nodes exceeds 1/3. However, the case Tarak describes here requires that the cluster has 2/3 byzantine participants. The probabilities of this are described by the [Hypergeometric distribution](https://en.wikipedia.org/wiki/Hypergeometric_distribution). With multiple hundred collectors in the mature protocol and cluster sizes around 50 nodes or above, the probability of reaching 2/3 byzantine fraction in a cluster is very small.  For example, consider 500 collectors in total, 166 of which are byzantine. If we partition the collectors in clusters of size 50, the probability that one particular cluster reaches 2/3 byzantine members is `2 * 10^-8`. One the one hand, we would sample constitute 10 cluster each epoch, but an epoch is also 1 week long. 

I think it would be nice if we could handle this edge case with little extra work. But I think we could pragmatically also just disregard this edge case as "vanishing probability".

jordanschalm replied:
>  it is reasonably likely to sample a cluster, whose local fraction of byzantine nodes exceeds 1/3. However, the case Tarak describes here requires that the cluster has 2/3 byzantine participants.

👍  I agree, I was conflating the "Byzantine-supermajority" case with "no honest supermajority" case - I'll resolve this
